### PR TITLE
add Allow header to 405 responses as required by RFC 7231

### PR DESCRIPTION
**Pull request by adriangb**
To be applied on top of #1435, replaces #1434

### CODE REVIEW

File: starlette/endpoints.py [hunk 1]
```diff
@@ -17,6 +17,10 @@ def __init__(self, scope: Scope, receive: Receive, send: Send) -> None:
         self.scope = scope
         self.receive = receive
         self.send = send
+        self._allowed_methods = filter(
+            lambda method: getattr(self, method.lower(), None) is not None,
+            ["GET", "HEAD", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"],
+        )
```
Comment (lovelydinosaur) [inline starlette/endpoints.py]: I *guess* I probably find a list comprehension style easier to understand...

```python
self._allowed_methods = [
    method for method in HTTP_METHODS
    if getattr(self, method.lower(), None) is not None
]
```

Butttttt it's just a matter of preference isn't it. Possibly useful feedback(?), but not a blocker to me.

**Review by lovelydinosaur — APPROVED** [review]
Fab, yup. One bit of feedback, but up to you.

File: starlette/endpoints.py [hunk 2]
```diff
@@ -17,10 +17,11 @@ def __init__(self, scope: Scope, receive: Receive, send: Send) -> None:
...
         self.send = send
-        self._allowed_methods = filter(
-            lambda method: getattr(self, method.lower(), None) is not None,
-            ["GET", "HEAD", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"],
-        )
+        self._allowed_methods = [
+            method
+            for method in ("GET", "HEAD", "POST", "PUT", "PATCH", "DELETE", "OPTIONS")
+            if getattr(self, method.lower(), None) is not None
+        ]
```
Comment (lovelydinosaur) [inline starlette/endpoints.py:24]: Super. 🌟

File: starlette/endpoints.py (hunk 1 above)
Comment (adriangb) [reply starlette/endpoints.py]: yep makes sense, thanks for the suggestion!

**Pull request merged**